File: solutions/python/say/1/say.py
def say(number):
    if number < 0 or number > 999_999_999_999:
        raise ValueError("input out of range")
    if number == 0:
        return "zero"
    chunks = {
        1_000_000_000: "billion",
        1_000_000: "million",
        1_000: "thousand"
    }
    result = ""
    for key, value in chunks.items():
        if number >= key:
            result += f"{get_chunk(number // key)} {value} "
            number %= key
    return result + get_chunk(number) if number > 0 else result[:-1]    
    
    
def get_chunk(number):
    basics = {
        1: "one",
        2: "two",
        3: "three",
        4: "four",
        5: "five",
        6: "six", 
        7: "seven",
        8: "eight",
        9: "nine",
        10: "ten",
        11: "eleven",
        12: "twelve",
        13: "thirteen",
        14: "fourteen",
        15: "fifteen",
        16: "sixteen",
        17: "seventeen",
        18: "eighteen",
        19: "nineteen"
    }
    tenths = {
        90: "ninety",
        80: "eighty",
        70: "seventy",
        60: "sixty",
        50: "fifty",
        40: "forty",
        30: "thirty",
        20: "twenty"    
    }
    result = ""
    if number >= 100:
        result += f"{basics[number // 100]} hundred"
        number %= 100
        result += ' ' if number > 0 else ''
    for key, value in tenths.items():
        if number >= key:
            number -= key
            result += f"{value}{'-' if number > 0 else ''}"
            break
    result += basics[number] if number > 0 else ''
    return result

File: solutions/javascript/say/1/say.js
//
// This is only a SKELETON file for the 'Say' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const say = (n) => {
  if (n < 0 || n > 999_999_999_999)
    throw new Error('Number must be between 0 and 999,999,999,999.');
  if (n === 0)
    return 'zero';
  
  const thousands = [
    [1_000_000_000, 'billion'],
    [1_000_000, 'million'],
    [1_000, 'thousand']
  ];

  let res = '';

  for (const [limit, label] of thousands) {
    if (n >= limit) {
      const chunk = Math.floor(n / limit);
      n -= chunk * limit;
      res += `${getChunk(chunk)} ${label} `;
    }
  }

  return n> 0 ? res + getChunk(n) : res.trim();
  };

const getChunk = (n) => {
  const ones = [
    '', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
  ];
  const tens = [
    '', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
  ];

  let res = '';
  
  if (n > 99) {
    const hund = Math.floor(n / 100);
    n -= hund * 100;
    res = `${ones[hund]} hundred`;
  }

  if (n === 0)
    return res;
  else
    res = res ? res + ' ' : res;

  if (n > 19) {
    const mod = n % 10;
    return res + tens[Math.floor(n / 10)] + (mod !== 0 ? `-${ones[mod]}` : '');
  } else 
    return res + ones[n];
}
